lib/node_modules/@stdlib/math/generics/random/sample/lib/factory.js

```diff
@@ -20,35 +25,119 @@ var slice = Array.prototype.slice;
 /**
 * Returns a function to sample elements from an array-like object.
 *
-* @param {PositiveInteger} seed - integer-valued seed
+* @param {ArrayLike} [pool] - array-like object to sample from
+* @param {Options} [options] - function options
+* @param {PositiveInteger} [options.seed] - integer-valued seed
+* @param {NonNegativeInteger} [options.size] - sample size
+* @param {boolean} [options.replace=true] - boolean indicating whether to sample with replacement
 * @returns {Function} function to sample elements from an array-like object
 *
 * @example
-* var sample = factory( 323 );
+* var sample = factory({
+*     'seed': 232
+* });
+* var out = sample( 'abcdefg' );
+* // [ 'g', 'd', 'g', 'f', 'c', 'e', 'f' ]
+*
+* @example
+* var sample = factory( [1,2,3,4,5,6], {
+*     'seed': 232
+*     'size': 2
+* });
+* var out = sample();
+* // returns [ 6, 4 ]
+*
+* out = sample()
+* // returns [ 6, 5 ]
+*
+* @example
+* var sample = factory( [1,2,3,4,5,6], {
+*     'seed': 474
+*     'size': 3
+*     'replace': false
+* });
+* var out = sample();
+* // returns [ 4, 3, 6 ]
+*
+* out = sample()
+* // returns [ 1, 5, 2 ]
+*
+* out = sample()
+* // throws <Error>
+*
+* @example
+* var sample = factory( [0,1], {
+*     'size': 2
+* });
+*
+* var out = sample();
+* // e.g., returns [ 1, 1 ]
+*
+* out = sample({
+*     size: 10
+* });
+* // e.g, returns [ 0, 1, 1, 1, 0, 1, 0, 0, 1, 1 ]
+*
+* @example
+* var sample = factory( [0,1], {
+*     'size': 2
+* });
+*
+* var out = sample();
+* // e.g., returns [ 1, 1 ]
+*
+* out = sample({
+*     'replace': false
+*     'size': 1
+* });
+* // returns [ 0 ] or [ 1 ]
+*
+* out = sample();
+* // throws <RangeError>
```

Comment:
Re: consistency. The current behavior is only consistent from one perspective. From the opposite perspective, the behavior is inconsistent. For instance,

``` javascript
var mysample = sample.factory( [0,1], {
    'replace': true, 
    'size': 1 
});

// mutates
var out = mysample({
    'replace': false
});
```

``` javascript
// Does not mutate:
var out = sample( [0,1], {
    'replace': true, 
    'size': 1 
});
```

Here, the behavior is inconsistent: I configured a factory to not mutate. By overriding, now I have mutated state. 

And in seeing this, I see nothing inconsistent with `factory` and `sample` where 

``` javascript
var mysample = sample.factory( [0,1], {
    'replace': true, 
    'size': 1 
});

var out = mysample({
    'replace': false
});
```
and

``` javascript
var out = sample( [0,1], {
    'replace': false, 
    'size': 1 
});
```

In both cases, we mutate only once and transiently. In neither case are we mutating state. The former example may be considered a curried version of the latter, but without changing the underlying state; or in other words, in the former, we validate `x` only once and can perform different types of sampling using the same source pool (including where I might want to sample without replacement).

So here's the more underlying issue. Sampling without replacement does not imply mutation (see how you setup FisherYates in `sample1` and `renormalizing`; the opposite is, however, true: mutation, in this case, does imply sampling without replacement). For the current implementation, mutation is an implicit side effect. My desire is that this side effect be made explicit and something which is actively opt-in.

To this end, I don't believe that an extended explanation suffices. The current interface is a leaky abstraction (implementation details are exposed to the user). My sense is that, in this case, only by being explicit can we be consistent.